### PR TITLE
Refactor utils interception webdriver to typescript

### PR DESCRIPTION
**Pull request by Fabianopb**
## Proposed changes

Refactoring `webdriverio/utils/interception/webdriver.js` to TypeScript.

## Types of changes

- [x] Refactor from JS to TS

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: packages/webdriverio/src/utils/interception/webdriver.ts [hunk 1]
```diff
@@ -8,6 +8,8 @@ import { ERROR_REASON } from '../../constants'
  * compliant backend.
  */
 export default class WebDriverInterception extends Interception {
+    mockId: number = NaN;
```
Comment (Fabianopb) [inline packages/webdriverio/src/utils/interception/webdriver.ts:11]: Should this be initialized somehow here or then should we check that `mockId !== undefined` before calling the methods that require mockId?

File: packages/webdriverio/src/utils/interception/webdriver.ts [hunk 2]
```diff
@@ -16,6 +18,7 @@ export default class WebDriverInterception extends Interception {
     /**
      * allows access to all requests made with given pattern
      */
+    // @ts-ignore
```
Comment (Fabianopb) [inline packages/webdriverio/src/utils/interception/webdriver.ts:21]: Couldn't figure out yet how to match with the `calls` type from the class `Interception` (should it match?)

Also this gives the error:
```
'calls' is defined as a property in class 'Interception', but is overridden here in 'WebDriverInterception' as an accessor.ts(2611)
```

File: packages/webdriverio/webdriverio.d.ts [hunk 3]
```diff
@@ -37,6 +44,12 @@ declare namespace WebdriverIO {
          * to this callback will be returned to the client.
          */
         executeAsync: (script: string | ((...arguments: any[]) => void), ...arguments: any[]) => Promise<any>;
+
+        mockCalls(mockId: number): Promise<void>;
+
+        mockRequest(url: string, filterOptions: MockFilterOptions): { mockId: number };
+
+        respondMock(mockId: number, payload: object): Promise<void>;
```
Comment (Fabianopb) [inline packages/webdriverio/webdriverio.d.ts:52]: Is this the right way to go, adding the missing types here? If yes, any tips how to document them?

Comment (Fabianopb) [reply packages/webdriverio/webdriverio.d.ts:52]: Hummm... looks like this is not the right approach? Some tests failed in another module... 🤔

**Comment by Fabianopb** [comment]
@christian-bromann if you could please provide a little guidance in this one! 🙏

File: packages/webdriver/webdriver.d.ts [hunk 4]
```diff
@@ -2504,21 +2504,21 @@ declare namespace WebDriver {
          * Receive request information about requests that match the mocked resource.
          * https://wiki.saucelabs.com/display/DOCS/Custom+Sauce+Labs+WebDriver+Extensions+for+Network+and+Log+Commands
          */
-        getMockCalls(mockId: string): ProtocolCommandResponse;
+        getMockCalls(mockId: number): ProtocolCommandResponse;
```
Comment (Fabianopb) [inline packages/webdriver/webdriver.d.ts:2507]: Tests require `number` so should this be changed?

File: packages/webdriverio/webdriverio.d.ts (hunk 3 above)
Comment (Fabianopb) [reply packages/webdriverio/webdriverio.d.ts]: Outdated, figured out already

**Comment by Fabianopb** [comment]
I can't add reviewers to this PR, perhaps @mgrybyk ?

File: packages/webdriver/webdriver.d.ts (hunk 4 above)
Comment (christian-bromann) [reply packages/webdriver/webdriver.d.ts:2507]: it should be a string if these commands have been defined as strings before

File: packages/webdriverio/src/utils/interception/webdriver.ts (hunk 1 above)
Comment (christian-bromann) [reply packages/webdriverio/src/utils/interception/webdriver.ts:11]: it should be defined as optional

File: packages/webdriver/webdriver.d.ts (hunk 4 above)
Comment (Fabianopb) [reply packages/webdriver/webdriver.d.ts:2507]: Then I guess all tests here have to be fixed to assert agains string, right?
https://github.com/webdriverio/webdriverio/blob/master/packages/webdriverio/tests/utils/interception/webdriver.test.js#L32

File: packages/webdriverio/src/utils/interception/webdriver.ts (hunk 1 above)
Comment (Fabianopb) [reply packages/webdriverio/src/utils/interception/webdriver.ts:11]: Then how do you suggest dealing with those?
image

File: packages/webdriver/webdriver.d.ts (hunk 4 above)
Comment (Fabianopb) [reply packages/webdriver/webdriver.d.ts]: Apparently not, I might have made some confusion while I was developing, changed to string now and everything worked

File: packages/webdriverio/src/utils/interception/webdriver.ts (hunk 1 above)
Comment (Fabianopb) [reply packages/webdriverio/src/utils/interception/webdriver.ts]: I could cast, maybe

**Comment by erwinheitzman** [comment]
Thank you for your contribution.

I prefer to fully migrate packages, not parts of packages.

The story is also to migrate the webdriverio package, so I'm not sure what to think of this PR at this moment.
I think migrating only parts of packages will make it chaotic and less obvious on what is and is not yet migrated, and it will make it hard for others to start working on things.

**Comment by christian-bromann** [comment]
@erwinheitzman it is actually the way I recommended it. Porting the whole package just creates a too large PR to review. It makes sense to me to find little portions to do it step by step. The interception logic seems like a good start.

**Comment by Fabianopb** [comment]
> @erwinheitzman it is actually the way I recommended it. Porting the whole package just creates a too large PR to review. It makes sense to me to find little portions to do it step by step. The interception logic seems like a good start.

I actually second this approach, I'm currently working in a huge repo where a few dozens of devs contribute and we're migrating it to TS. The only feasible way so far has been to do it step by step! :)

Also, as someone who's still getting acquainted with webdriverio project, I feel more confident to do it in small steps!

**Review by christian-bromann — CHANGES_REQUESTED** [review]
The implementation of this file is not quite correct (it is actually also not yet supported). All `browser.mockCalls(...)` calls need to be `browser.getMockCalls(...)`. The command is implemented in the Sauce Labs WebDriver protocol.

**Comment by erwinheitzman** [comment]
@Fabianopb first of all my apologies if it felt like I was criticizing you, I am very glad to have you contribute to the project! ❤️

I thought it might come with some risks but seems like it's actually rather easy to migrate like this so I'm all for it, thanks 👍

**Comment by Fabianopb** [comment]
> @Fabianopb first of all my apologies if it felt like I was criticizing you, I am very glad to have you contribute to the project! ❤️
> 
> I thought it might come with some risks but seems like it's actually rather easy to migrate like this so I'm all for it, thanks 👍

No biggie! I just wanted to share my two cents about own experience in a similar case! :)

**Comment by Fabianopb** [comment]
> The implementation of this file is not quite correct (it is actually also not yet supported). All `browser.mockCalls(...)` calls need to be `browser.getMockCalls(...)`. The command is implemented in the Sauce Labs WebDriver protocol.

Understood, I have now fixed the implementation and related test.

**Review by christian-bromann — APPROVED** [review]
Would be nice to find a way to remove the `@ts-ignore` at some point but let's move forward, the rest looks good 👍

**Pull request merged**